### ISSUE TITLE
Make rules generic (to reenable `-c` flag as well).

### ISSUE THREAD

**Issue by angerman** [snowleopard/hadrian#538]
As noted in [this comment](https://github.com/snowleopard/hadrian/pull/531#issuecomment-377258787), we currently generate specific rules for all known libraries. We would be better off to just build generic rules and then parse the details from the path.

That resolves on the one hand the need to know everything before hand to generate all specific rules, and I would expect this to also speedup the startup process of hadrian.

So what do we need? We'll need functions to extract at least:
 - stage
 - way
 - package

from a path.

---
### Example:
right now we generate
```haskell
    root -/- libDir context -/- pkgId -/- archive way pkgId %> \_ -> do
```
where `root` is a `FilePath` computed from `buildRootRules`, and we also have:
```haskell
libDir :: Context -> FilePath
libDir Context {..} = stageString stage -/- "lib"

archive :: Way -> String -> String
archive way pkgId = "libHS" ++ pkgId ++ (waySuffix way <.> "a")
```

As such we basically want to match something like:
```
${root}/${stage}/lib/${pkgId}/libHS${pkgId}${waySuffix}.a
```
And use a pattern like
```haskell
root -/- "*.a" %>\a -> do
  ...
```
here. And then extract the details from `a` and `fail` if it's not a match.

#### Additional notes
maybe we want to reuse some URL-matcher/router from one of the web-frameworks to extract the details easily and not write the logic by hand, while using an already performance tuned library?

**Comment by snowleopard** [comment]
@angerman Awesome, thank you very much for recording your thoughts here.

**Comment by alpmestan** [comment]
@snowleopard Would you be okay with defining simple parsers for this using `parsec`, given that it's now a dependency of `Cabal` and is therefore already a (transitive) dependency ?

**Comment by alpmestan** [comment]
@snowleopard I was actually just speaking about `parsec` itself, not Cabal's parsers. We don't need something terribly fancy, and we do have to map things to our own types anyway (`Way`, `Stage`, `Package`, `Context` and so on)

**Comment by snowleopard** [comment]
Ah, I see. Well, depending on one more GHC library seems like an unnecessary high cost, both complexity wise and compilation time wise -- there are two dozen extra modules to compile!

If all we want is to parse `${root}/${stage}/lib/${pkgId}/libHS${pkgId}${waySuffix}.a`, then [splitPath](https://hackage.haskell.org/package/filepath-1.4.2/docs/System-FilePath-Posix.html#v:splitPath) and [stripPrefix](https://hackage.haskell.org/package/base-4.11.0.0/docs/Data-List.html#v:stripPrefix) are sufficient and will do the job at much lower cost.

**Comment by alpmestan** [comment]
Well, we already depend on Cabal, so we already build parsec whenever we build hadrian now. If you however still would like to avoid the complexity, I'll do without parsec, using something like what you suggest.

**Comment by snowleopard** [comment]
> Well, we already depend on Cabal, so we already build parsec whenever we build hadrian now

@alpmestan Ah, indeed -- that's a good point. So, you are saying that the cost of using `parsec` is actually zero? We already transitively depend on it, and we already build it. In this case, let's give `parsec` a try 👍

**Comment by alpmestan** [comment]
@snowleopard Exactly, the cost is zero, but the usability/readability of the code will be greater IMO. I'm preparing a proof-of-concept patch with parsec that pretty much parses the `Context` out of filepaths, at least just for `

**Comment by snowleopard** [comment]
@alpmestan Thanks, sounds good! Sorry for being slow today :)

**Comment by alpmestan** [comment]
@snowleopard @angerman I have a WIP branch at https://github.com/snowleopard/hadrian/compare/master...alpmestan:generic-library-rules that implements Moritz's suggestion. I still seem to be able to build a GHC just fine if I run the `./boot && ./configure` step myself explicitly, but not if I start from a clean source tree and run hadrian with `-c`. The error is:

```
[hadrian/cfg/system.cfg, settings, mk/config.h]
about to need ["hadrian/cfg/system.config.in","settings.in","mk/config.h.in"]
about to run ./configure
in doWith
builder ready
shakeArgsWith                      0.000s    0%                           
Function shake                     0.272s   65%  =========================
Database read                      0.001s    0%                           
With database                      0.000s    0%                           
Running rules                      0.140s   33%  ============             
Pool finished (14 threads, 4 max)  0.000s    0%                           
Total                              0.414s  100%                           
Build system error - indirect recursion detected:
  Key value 1:  OracleQ (KeyValue ("hadrian/cfg/system.config","host-os"))
  Key value 2:  hadrian/cfg/system.config
  Key value 3:  hadrian/cfg/system.config settings mk/config.h
  Key value 4:  OracleQ (KeyValue ("hadrian/cfg/system.config","project-version"))
Rules may not be recursive
```

(the first few lines come from debugging output I added -- see my branch)

Do any of you have any idea of where that cycle comes from and why my patch makes this happen? I won't have a lot of time to look into this for the end of this week, just a little bit, so any help would be appreciated =)

**Comment by izgzhen** [comment]
@alpmestan regarding the `-c` failure above, I noticed this when I was debugging the Circle CI error, and it appears even in the `selftest` target. I don't know much else though...

**Comment by alpmestan** [comment]
My patch got submitted and merged in #571. Shall we close this @angerman?

**Comment by angerman** [comment]
Sounds good!
